### DISCUSSION TITLE
Intermittent Connection attempt timed out error

### DISCUSSION THREAD

**Discussion by aviralsingh21** [dbeaver/dbeaver#34289] [Q&A] [answered]
I have a docker swarm architecture for 3 PostgreSQL (Patroni + ETCD), 3 Pgbouncer and 1 HAProxy.
Where, Client/Application will request HAProxy to connect with database. Then HAProxy will connect to Pgbouncer and Pgbouncer will route connection to Primary database.

After successful setup now I am trying to establish a connection in DBeaver by mentioning the connections details for haproxy, I am experiencing a random behaviour. Although using psql cli utility I am able to make a successful connection.
At first few tries, I get the _Connection attempt timed out error_ but when I restart the haproxy service few times (3 to 4 times) then connections is established. Or if I leave haproxy service keep running for an hour or 2 then also connection is established.
I have been experiencing this issue daily., since the time of setup. 

Funny part is, I have 2 users: postgres and test. And 2 databases with same name as users: postgres and test.
So, when I try to test a connection between:
**User  ::  Database  ::  Connection Status**
postgres  ::  postgres  ::  **Success**
postgres  ::  test  ::  **Success**
test  ::  postgres  ::  **Success**
test  ::  test  ::  **Sometimes Success, Sometimes _Connection attempt timed out_ error**.

At the time of Error:
![PkCNjZ8Uyn](https://github.com/dbeaver/dbeaver/assets/91733607/6b5b80b9-ee78-4270-bc4a-ad0105f5e9c7)

At successful connection:
![ji8BWQxOMv](https://github.com/dbeaver/dbeaver/assets/91733607/c90143cd-fc78-4df2-9e1d-776aa111d134)

**Comment by TB-HunBoka** [comment]
I never used PostgreSQL but I had similar issue with the oracle driver. The default timeout dimension supposed to be secundum, but for some reason, there were an oracle driver that used milliseconds, so I got instant timeout on connection.

Do you know how much time do you wait to connect when you got the timeout error?
Can you check if it is a regular server-side timeout configured in the PostgreSQL Driver properties loginTimeout and connectionTimeout settings (or in the psqlrc.conf)?

Also, is there any additional info in the details?

**Reply by aviralsingh21** [reply]
Hi @TB-HunBoka,

1. At the time of connection attempt timed out error, it is taking around 20 seconds to give error. But when I keep running services for at least 2 hour, then connection is made within 10 seconds.

2. I checked in the DBeaver Driver settings,  loginTimeout and connectionTimeout values are 20 seconds for both. And timout settings in pgbouncer and database is disabled.

3. No, there is no extra info in details.

**Reply by TB-HunBoka** [reply]
Hi @aviralsingh21,

Looks like you exceed the default 20 seconds limit, you can change these settings as you mentioned, no server-side timeout is defined, so the driver drops the login attempt. Check it with 240 seconds and see how long you need to wait until the connection established.

There is something in your environment that creates huge delay for you. You can verify this by running the smallest possible SQL query and check the response time or check the HAproxy logs for the login attempt.

**Reply by aviralsingh21** [reply]
Hi @TB-HunBoka,

Thank You for your inputs.

1. I tried with loginTimeout of 240 seconds. Still got the time out error.
2. As you suggested, I ran the `select inet_server_addr();` simple query in database accessing haproxy using psql utility. It took 4.059 ms at first and 0.841 ms for second time to execute.

3. There is not much in haproxy logs. Below is the complete logs of HAProxy:

> [NOTICE]   (1) : New worker (8) forked,
> [NOTICE]   (1) : Loading success.,
> [WARNING]  (8) : Server postgresql/postgresqldb-1 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 6ms. 2 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.,
> [WARNING]  (8) : Server postgresql/postgresqldb-3 is DOWN, reason: Layer7 wrong status, code: 503, info: "Service Unavailable", check duration: 1ms. 1 active and 0 backup servers left. 0 sessions active, 0 requeued, 0 remaining in queue.

I suppose Service Unavailable warning is because I have configured haproxy to connect to node with primary role only. And for now postgresqldb-2 is primary.

**Reply by TB-HunBoka** [reply]
Hello @aviralsingh21,

I wonder why you have more than 20 seconds login then.
However, as a last chance you can change not only the loginTimeout but also the connectionTimeout to 240 seconds. Both settings can cause a timeout error.

**Reply by aviralsingh21** [reply, marked as answer]
Hello @TB-HunBoka,
It seems, it was the issue with pgbouncer pool_size value. When I increased the value of pool_size, it was able to connect instantly.

**Reply by TB-HunBoka** [reply]
Hello @aviralsingh21,

I am happy you found a solution.
Thats a strange issue. My understanding was that you have only one or two active connections at a time.

I suggest accepting your comment as correct answer, so others with similar issue can jump to the fix.
